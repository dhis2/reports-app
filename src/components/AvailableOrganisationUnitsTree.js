import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import i18n from '@dhis2/d2-i18n'
import { OrgUnitTreeMultipleRoots } from '@dhis2/d2-ui-org-unit-tree'
import { selectOrganisationUnit } from '../redux/actions/organisationUnits'

export function AvailableOrganisationUnitsTree({
    selectOrganisationUnit,
    loading,
    collection,
    selected,
}) {
    if (loading) {
        return <span>{i18n.t('Updating Organisation Units Tree...')}</span>
    }

    return (
        <div className="org-unit-tree-wrapper">
            <OrgUnitTreeMultipleRoots
                hideMemberCount
                roots={collection}
                selected={selected ? [selected.path] : []}
                initiallyExpanded={collection.map(unit => `/${unit.id}`)}
                onSelectClick={selectOrganisationUnit}
            />
            <style jsx>{`
                div {
                    border: 1px solid #bcbcbc;
                    overflow: auto;
                    width: 100%;
                    box-sizing: border-box;
                    font-weight: 300;
                    max-height: 400px;
                    padding: 10px 5px;
                }
            `}</style>
        </div>
    )
}

AvailableOrganisationUnitsTree.propTypes = {
    collection: PropTypes.array.isRequired,
    loading: PropTypes.bool.isRequired,
    selectOrganisationUnit: PropTypes.func.isRequired,
    selected: PropTypes.object,
}

const mapStateToProps = ({ organisationUnits }) => ({
    loading: organisationUnits.loading,
    collection: organisationUnits.collection,
    selected: organisationUnits.selected,
})

export default connect(mapStateToProps, { selectOrganisationUnit })(
    AvailableOrganisationUnitsTree
)
