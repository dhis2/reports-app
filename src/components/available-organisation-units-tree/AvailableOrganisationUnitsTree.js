/* React */
import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

/* d2-ui components */
import { OrgUnitTree } from '@dhis2/d2-ui-org-unit-tree'

/* App context */
import AppContext from '../../pages/AppContext'

/* i18n */
import i18n from '../../utils/i18n/locales'
import { i18nKeys } from '../../utils/i18n/i18nKeys'

/* styles */
const defaultStyles = {
    tree: {
        border: 'solid 1px #bcbcbc',
        overflow: 'auto',
        width: '100%',
    },
}

export class AvailableOrganisationUnitsTree extends PureComponent {
    static propTypes = {
        d2: PropTypes.object.isRequired,
        onChange: PropTypes.func,
        style: PropTypes.object,
    }

    static defaultProps = {
        onChange: null,
        style: defaultStyles,
    }

    constructor() {
        super()

        this.state = {
            selected: [],
            rootWithMember: null,
        }
    }

    componentDidMount() {
        if (this.state.rootWithMember == null) {
            this.props.d2.models.organisationUnits
                .list({
                    paging: false,
                    level: 1,
                    fields:
                        'id,displayName,path,children::isNotEmpty,memberCount',
                })
                .then(organisationUnitsResponse => {
                    const organisationUnits = organisationUnitsResponse.toArray()
                    this.setState({
                        rootWithMembers: organisationUnits[0],
                    })
                })
                .catch(() => {
                    // TODO manage error
                })
        }
    }

    handleOrgUnitClick = (event, orgUnit) => {
        if (!this.state.selected.includes(orgUnit.path)) {
            this.setState({ selected: [orgUnit.path] })
            if (this.props.onChange) {
                const selectedOrganisationUnitSplitted = orgUnit.path.split('/')
                const selectedOrganisationUnitId =
                    selectedOrganisationUnitSplitted[
                        selectedOrganisationUnitSplitted.length - 1
                    ]
                this.props.onChange(selectedOrganisationUnitId)
            }
        }
    }

    render() {
        if (this.state.rootWithMembers) {
            return (
                <div style={this.props.style.tree}>
                    <OrgUnitTree
                        hideMemberCount
                        root={this.state.rootWithMembers}
                        selected={this.state.selected}
                        initiallyExpanded={[
                            `/${this.state.rootWithMembers.id}`,
                        ]}
                        onSelectClick={this.handleOrgUnitClick}
                    />
                </div>
            )
        }

        return (
            <span>
                {i18n.t(
                    i18nKeys.availableOrganisationUnitsTree.updatingMessage
                )}
            </span>
        )
    }
}

export default props => (
    <AppContext.Consumer>
        {appContext => (
            <AvailableOrganisationUnitsTree d2={appContext.d2} {...props} />
        )}
    </AppContext.Consumer>
)
