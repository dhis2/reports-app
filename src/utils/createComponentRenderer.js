import React from 'react'
import { shallow } from 'enzyme'

export default function createComponentRenderer(Component, defaultProps) {
    return function(customProps) {
        const props = { ...defaultProps, ...customProps }
        return shallow(<Component {...props} />)
    }
}
