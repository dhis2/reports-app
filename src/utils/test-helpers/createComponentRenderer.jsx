import { shallow } from 'enzyme'
import React from 'react'

export default function createComponentRenderer(Component, defaultProps) {
    return function (customProps) {
        const props = { ...defaultProps, ...customProps }
        return shallow(<Component {...props} />)
    }
}
