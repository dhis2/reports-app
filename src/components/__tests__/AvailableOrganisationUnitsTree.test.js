import createComponentRenderer from '../../utils/test-helpers/createComponentRenderer'

/* App components */
import { AvailableOrganisationUnitsTree } from '../AvailableOrganisationUnitsTree'

jest.mock('@dhis2/d2-ui-org-unit-tree', () => ({
    OrgUnitTreeMultipleRoots: 'OrgUnitTreeMultipleRoots',
}))

describe('Test <OrganisationUnitGroupSets /> rendering:', () => {
    const selectMock = jest.fn()

    const defaultProps = {
        selectOrganisationUnit: selectMock,
        ready: true,
        loadingError: '',
        collection: [],
        selected: null,
    }

    const componentRenderer = createComponentRenderer(
        AvailableOrganisationUnitsTree,
        defaultProps
    )

    afterEach(() => {
        selectMock.mockClear()
    })

    it('Should render OrgUnitTree when ready', () => {
        expect(componentRenderer()).toMatchSnapshot()
    })

    it('Should render a loading message when not ready', () => {
        expect(componentRenderer({ ready: false })).toMatchSnapshot()
    })

    it('Should render an error message when it has a loadingError', () => {
        expect(componentRenderer({ loadingError: 'Oops' })).toMatchSnapshot()
    })
})
