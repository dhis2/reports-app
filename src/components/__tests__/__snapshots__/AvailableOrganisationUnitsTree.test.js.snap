// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <OrganisationUnitGroupSets /> rendering: Should render OrgUnitTree when ready 1`] = `
<div
  style={
    Object {
      "border": "solid 1px #bcbcbc",
      "overflow": "auto",
      "width": "100%",
    }
  }
>
  <OrgUnitTreeMultipleRoots
    hideMemberCount={true}
    initiallyExpanded={Array []}
    onSelectClick={[Function]}
    roots={Array []}
    selected={Array []}
  />
</div>
`;

exports[`Test <OrganisationUnitGroupSets /> rendering: Should render a loading message when not ready 1`] = `
<span>
  Updating Organisation Units Tree...
</span>
`;

exports[`Test <OrganisationUnitGroupSets /> rendering: Should render an error message when it has a loadingError 1`] = `
<span
  style={
    Object {
      "color": "#f44336",
    }
  }
>
  Oops
</span>
`;
